_posts/2020-04-22-#17994.md

```diff
@@ -0,0 +1,54 @@
+---
+layout: pr
+date: 2020-04-22
+title: "Flush undo files after last block write"
+pr: 17994
+authors: [kallewoof]
+components: ["validation"]
+host: vasild
+status: upcoming
+commit:
+---
+
+## Notes
+
+In this Review Club we will dive into how blocks are downloaded and stored on disk. How disk access is arranged and a bug it has caused.
+
+Some concepts that would help to grasp what's going on:
+
+- [Initial Block Download](https://bitcoin.org/en/glossary/initial-block-download)
+
+- "block undo" - all the information that is necessary to revert a block, if a chain reorganization happens, for example.
+
+- [class FlatFileSeq](https://doxygen.bitcoincore.org/class_flat_file_seq.html)
+
+- file [flush](https://linux.die.net/man/3/fflush) - moving all buffered data from the `FILE*` buffers to the OS (i.e. in kernel buffers). It may not necessary hit the disk yet.
+
+- file [fsync](https://linux.die.net/man/2/fsync) - moving all modified data from the OS buffers to the disk.
+
+## Questions
+
+### Overview
+
+- How/when is the blocks database (blocks/{blk,rev}\*.dat) created?
+- Can we create an undo for a given block without having all prior blocks?
+- Do we ever modify existent data in the blocks database or we just append new blocks and their undos?
+
+### Dive
+
+- How is space allocated in the files when new data is appended? Why?
+- What does to "finalize" a file mean?
+- What is the bug that the PR is fixing?
+- How would you demonstrate the bug existence?
```

Comment:
How would you reproduce the bug or show it exists?

Reply:
Done